### ISSUE TITLE
Issue after updating to npm 3.x

### ISSUE THREAD

**Issue by wesleytodd** [timoxley/linklocal#20]
We were previously using linklocal and bulk to install some local packages.  While trying to upgrade to npm 3.x, the flattening of the dep tree is causing linked packages to NOT have access to the shared deps.  Atleast that is how I have diagnosed the problem.

Although I do not think this is a direct bug with linklocal, it seems like I should start here rather than creating noise on the npm repo, as they have plenty to deal with already.  Do you have a suggested solution to solve this new problem?

NOTE: feel free to ask for an actual outline of our use case if you need it and I will write up more details.

**Comment by timoxley** [comment]
I haven't looked at the npm 3 problem since they released, will need to investigate. Please tell me more about your usecase

**Comment by wesleytodd** [comment]
While trying to make a reduced test case to show you I am seeing some different behavior.  I will close this and re-open it if I can narrow it down more.

**Comment by timoxley** [comment]
@wesleytodd any updates?

**Comment by wesleytodd** [comment]
I had not had a chance to look again, some prod issues had to be sorted out on our product.  But I will work on this a bit today and post back with my findings!  sorry for the delay!

**Comment by wesleytodd** [comment]
Ok, so it appears that my original issue was with some modules previously having been installed with npm 2.x being used.  So the original issue is not valid.  BUT....

It is fairly clear to me that local deps are not actually viable right now.  The odd semantics of how modules do their lookups when linked creates some downright horrible conditions where it becomes almost impossible to tell where it should be loading code from.  I think this deserves some thought and would be happy to participate in a larger discussion on this feature and how to make it better.  But only if other people would use it, which personally I wouldn't if I had a choice in the matter...

**Comment by wesleytodd** [comment]
Oh, and here is my test case: https://github.com/wesleytodd/npm-linklocal-test

**Comment by yoshuawuyts** [comment]
Yeah, so the way I work around this is by never providing `../` paths in package.json. What I settled on for my own projects is to use a flat directory structure:
```sh
./package.json         # specify ALL linklocal dependencies
./app-1/package.json
./app-2/package.json
./dep-1/package.json
./dep-2/package.json
```
Though this may not be what you're currently doing (or particularly like) I've found it is the easiest to reason about and the least prone to errors.

**Comment by wesleytodd** [comment]
@yoshuawuyts Not completely a fan of the flat structure that puts apps on the same level as packages, but I can see the usefulness.  

The problem is that our apps **need** inter-dependency dependencies *(yeah I had to re-read that statement to lol)*.  We have experimented with publishing them with MAJOR pains, (read: developer forgetting to update the package.json before trying to push to production on a feature that is 2-3 weeks in the making without that developer in the office at the time of releasing, good luck figuring out which module is missing its updates....FML).

If this was a new application it would be feasible to make changes to the structure, sadly that is not the case.  And, linklocal has been working pretty well for us in this case with our existing structure.
